### ISSUE TITLE
wiki quickie

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#52]
In order to finish off the readme (which will be shorter than it currently is), I need to first have ~~three~~ SIX wiki pages created (because the readme [[ed: or other wiki pages]] will point to them). They are

* [x] Overview
* [x] Backup
* [x] Implementation
* [x] Multiple Profiles
* [x] Multiple Profiles [Firefox Portable]
* [x] Concurrent Profiles

Overview: like what is a user.js, why have one (migration, enforcement etc), thats its editable in a text editor, how it works (process - case sensitive, what happens with syntax errors, what happens with type mismatches, the relationship between user.js->prefs.js->about:config and that it is a ff start that uses it), and how to use it (as in it's one way - so comment out in js, reset in about:config ) etc. How comment // works etc.

Backup - basically where to find your profile folder and to back that up. backing up prefs is not enough. There are other objects that can be changed, such as sqlite files.

Implementation: where to put the user.js. Either use in a new profile, and how to switch to it. Or how to grab a portable etc. Also to read it and make changes before jumping in. That kind of thing.

The next page on troubleshooting is all done (but I could improve it later).

That's all. If anyone wants to draft one or all of these up, please do. I've been ignoring doing this for a week or so (semi-procrastinating), especially as I wanted to keep the issues down (closed 28, yay!). That and the sexy new headers and forms sections are awesome. We've been here 3 weeks, and done a LOT, seriously, a LOT - well done. These pages won't take that long, but I'm kinda dreading writing them all myself and from scratch. I would love some help. Just paste in here or pastebin, and I can format it.

**Comment by Roman-Nopantski** [comment]
Oh yeah. I don't need or want plagiarized copies, and I don't really want tons of links to what other people have. We should all know this stuff, and be able to write our own original version. Ideas on what to add are good (and we can always do a revision later). PK's has a section on different OS on where to find the profile folder. But we can do that from within FF as well. So anything typed up off the top of your head to get me/us started would be excellent.

**Comment by 2glops** [comment]
HI,
Just a rough thing:
 
Find and backup your Firefox profile
Launch Firefox then follow the path : Open menu > Open Help Menu > Troubleshooting Information, 
search for Profile Directory then click on "Open Directory",
your current profile will be opened in your file manager.
The profile name is <alphanumerics>.<profilename>, e.g. fg8ets6y.default . Note that prefs.js and user.js stands there.
Backup the entire profile in a safe location.

**Comment by Roman-Nopantski** [comment]
Thanks @2glops - FYI: go to about:profiles .. there is a "open folder" button right there. I don't know of an easier or faster way

**Comment by 2glops** [comment]
Right, as you wish.
But, from about:profiles it could be easy to get confused and not saving the right profile. You have to select the Root Directory folder of the profile in use.
From the troubleshooting information page, there is no such risk. And the page list a lot of useful informations about the profile in use, including Important Locked Preferences.

**Comment by Roman-Nopantski** [comment]
@2glops Yep, it's all good stuff. I haven't written it yet :) I was just showing you a faster way

**Comment by atomGit** [comment]
if it helps, you guys can copy/paste whatever you want from my stuff [here](http://12bytes.org/articles/tech/firefoxgecko-configuration-guide-for-privacy-and-performance-buffs) - no credit or link-back needed

**Comment by earthlng** [comment]
I agree. We shouldn't re-invent the wheel. If someone has already done the work, why shouldn't we get some "inspiration" from that ;)

**Comment by Roman-Nopantski** [comment]
Yeah, look, this stuff is documented and written about a thousand times - I'm just not going to copypasta ad verbatim, plus we can do better than most of them (go read a dozen and make sure to cover everything to get the best combined info). I opened this issue, because it would be nice to finish it off and the readme, so all we have to focus on is the js.

I have a timetable in mind (for martin's article about us which will include the move, the big changes, and future plans). I also want a little space between that article (which will drive a LOT of people here) and the 6 weekly stable changelog articles. FF53 lands 18 April, so maybe about a week prior for the article at latest (that's 4 whole weeks away), and then 2 weeks later, a week after stable lands, comes our first "changelog" .. and then 6 weekly after that. If the article is even earlier, then the better.

I can probably whip up the three wikis and the readme in one lazy Saturday/Sunday arvo, review them the next day etc. So it's not a major. For some reason in my head I am procrastinating over it. It's like the last thing I want to do (and has been for weeks). I'm so not into writing them right now, I'll drag this on for a couple of weeks :)

If I don't write it (*hint hint*), if I am just an editor and review it (and we can all review it), then I don't really care - I only care about it's quality. These are straight-forward technical/how-tos, so there's not much to get wrong or indeed include. Take @atomGit 's bits as a starting base (we have his blessing) and GIT to it

**Comment by earthlng** [comment]
Re: 2.2. Lock Prefs - why don't we link to http://kb.mozillazine.org/Locking_preferences ??
I mean I could copy it all into a Wiki page but I don't see the need for that.
There's no need to dumb this down (with pics etc) since only "advanced users" would/should lock prefs anyway and they should be able to read and understand all the necessary things to do from that mozillazine.

**Comment by Roman-Nopantski** [comment]
I haven't really gotten round to that section. And I'm going to need to leave some numbers open for some ideas I have. Using numbers so the sidebar orders nicely (there are other solutions to that though). I just threw that section 2 and some subsection titles as a filler.

My plan is (to keep the readme short) to link to the section 1's etc. That's stage 1.

Later, after we branch to lite-medium-hardened, then I want to set up running multiple concurrent profile info wikis etc - with recommendations to set different themes to easily recognise what profile you're in. And of course lock_prefs is on the cards as well. But first things first.

**Comment by earthlng** [comment]
Ok, I'll link 2.2 to that mozillazine for now. We can always change it later if necessary.

**Comment by Roman-Nopantski** [comment]
@earthlng I'm using this issue just to get those 3 pages done, cuz you seem antsy about them xD .. when this is closed, I will be creating a "sticky wiki"

**Comment by Roman-Nopantski** [comment]
@earthlng Ahh, I see what u did .. that mozillazine is horrible .. this is the one: https://developer.mozilla.org/en-US/Firefox/Enterprise_deployment

**Comment by Thorin-Oakenpants** [comment]
@earthlng https://github.com/ghacksuserjs/ghacks-user.js/wiki/3.1-Greasemonkey

**Comment by Thorin-Oakenpants** [comment]
also chucked this up : https://github.com/ghacksuserjs/ghacks-user.js/wiki/3.3-uBlock-Origin . there's also pages for noscript and umatrix

**Comment by earthlng** [comment]
Very nice. But as always I hate the formatting ;)
* I'd move the "You can test..." lines for the 1st and 3rd script on a newline.
* The 2nd script could use some fewer TABs.
* 2nd title should be `Conceal window.name` without the 's'
* 1st script has a single `"` before history.length in description

I was expecting some explanation for Greasemonkey based on the title, IMO we should name it UserScripts.
And maybe add a recommendation to use Greasemonkey. I member you once said something on ghacks about Tampermonkey, but Idk what that was about.
Also we can use the Anchors to link to scripts directly in the user.js fe. [Clear window.opener](https://github.com/ghacksuserjs/ghacks-user.js/wiki/3.1-Greasemonkey#clear-windowopener)

**Comment by earthlng** [comment]
ps. Idk about that regex line, maybe it'd be better to add an `@exclusion` line for `https://www.google.com/recaptcha/` - probably needs some testing. I think pastebin has a captcha, at least it always bothered me with one.

**Comment by Thorin-Oakenpants** [comment]
go for it man - edit as you like. You're the code monkey :)

**Comment by earthlng** [comment]
I obviously use it differently than you, but for what it's worth, here are mine:

```
no-remote-fonts: * true
* * * noop
behind-the-scene * 3p noop
behind-the-scene * 3p-frame noop
```

**Comment by Thorin-Oakenpants** [comment]
This is where you can add some insight. The uBo stuff, such as no-remote fonts helped me. See, I don;t even know the diff that "behind-the-scene" makes. So we can rather than add big sets, add separate lines like the font one and explain what they are

**Comment by earthlng** [comment]
> I was expecting some explanation for Greasemonkey based on the title, IMO we should name it UserScripts.

nvm, it's under Scripts and then Greasemonkey. It's fine

**Comment by Thorin-Oakenpants** [comment]
nah, good stuff man .. changed the name to User Scripts. But i'm not sure on what to put for a short intro. Greasemonkey for sure, but what else - tampermonkey?

**Comment by Thorin-Oakenpants** [comment]
>Very nice. But as always I hate the formatting ;)
 I'd move the "You can test..." lines for the 1st and 3rd script on a newline.
 The 2nd script could use some fewer TABs.
 2nd title should be Conceal window.name without the 's'
 1st script has a single " before history.length in description

can't move to a new line, you have to use a new paragraph :-(
title fixed
extra " fixed
will try to fix tabs but i'll prolly get it all wrong

**Comment by 2glops** [comment]
My 2 cents for implementing:

Clone or download the user.js at :
https://github.com/ghacksuserjs/ghacks-user.js

First of all, read carefully the user.js and the documentation. Some site breakage could or will happen.
Second, make changes to the user.js according to your needs and expectations. You can change values or comment out any pref.

Backup and find you profile-s :
https://github.com/ghacksuserjs/ghacks-user.js/wiki/1.2-Backing-Up
You can also create a new profile by lauching the profile manager with the command :
firefox -P

Now, be sure Firefox is closed and put the user.js file in your Firefox profile directory. 

Re-start Firefox, that's all.
You can now check your main or most visited sites to identify if there is any breakage.

**Comment by Thorin-Oakenpants** [comment]
I already wrote the backup, all off the top of my head - as is my draft for the overview page and implementing (and it's that page that has the create new profile info, where to put the user.js etc).

I still think the best way for users is to jump right in (after reading the [warning] and [setup] tags as well as the three items specifically mentioned in the header section) - Create a new profile and then copy their existing profile into it - basically to clone it. Or just back it up. This way they have all their bookmarks and cookies and extensions all ready to go - etc which would be easier for them to use/test with. SFA breaks anyway, seriously, it like 5 prefs max (indexeddb, workers) and some custom ones like what to clear, urlbar/searchbar, session restore, maybe plugins - nothing really. There's nothing harmful in the js that can't be tweaked to sort out site breakage or lost functionality, and as for losing anything - lost history is no big deal (you can regenerate more!) and they were warned, so I'm not counting that. It's not like we wipe people's passwords - its a pretty tight but generous js IMO. Once we land branches then people will have even less to think about.

**Comment by earthlng** [comment]
Nice job on the backup page, Pants. It's a bit excessive for my taste but I guess that makes sense if someone needs a tutorial on how to backup their profile. some nits if you don't mind:
* wouldn't the first paragraph fit better under the Implementation?
* the 3rd paragraph is a bit confusing and could be dropped completely. Why make it more complicated - tell them to backup the whole profile (as you did) and don't even mention the prefs.js.
fe. "if you know what you're doing, you can do otherwise" - if they are reading that page they probably don't know what they're doing

We should try to keep everything as short and relevant as possible. for example

> Alternatively you can use the menu item ....

Isn't one way to find the profile directory enough?

> The next wiki section, Implementation, outlines some methods you could use to test out the user.js and preferences, such as using your existing profile, or a copy of it, or a new one and whether to implement piecemeal or whole hog

I would delete everything after `test out the user.js and preferences` because if it's outlined in the next section you don't need to already mention it in this section.

**Comment by Thorin-Oakenpants** [comment]
Well, got to use simple/layman terms if possible, use pictures, lower the bar and all that - I hate to say that, but it's a reality. It's also why I'm not going to quantify statements now (after the edit I just did) such as just using pref.js or rolling back/migrating specific data files. We can just recommend ONE way. If someone is an expert, good on them.

Also, think of the 4 pages like a single article - linkage previous at start, link next at end. It's not that bad.

- First paragraph rams home the profound effect it can have - thus to backup - warning & seque in one. *I have shortened the paragraph*
- i wanted to mention that prefs.js is not enough as a backup. because a lot of places suggest that (if u know what you;re doing it can be done no worries, but I'm not going to publicly endorse it). But yeah, TMI. *Have merged those paragraphs and trimmed the fat*
- added the about:support page info because @2glops mentioned it above, reason: it only shows the CURRENT profile (so users can't screw up). That said the about:profiles also says which is the profile in use... (3 minutes later)  .. yeah, kinda looks cleaner, but man, we're getting down to twitter lengths...
- trimmed last paragraph

check it out

**Comment by Thorin-Oakenpants** [comment]
draft up: https://github.com/ghacksuserjs/ghacks-user.js/wiki/1.1-Overview . Yes its long, but it covers a lot. It's actually very succinct for what it covers and I don't think anything else comes near to it on the web. All written off the top of my head.

**Comment by Atavic** [comment]
[Section 3.3](https://github.com/ghacksuserjs/ghacks-user.js/wiki/3.3-uBlock-Origin) could give a preset mode for uBlock, are you suggesting [Hard Mode](https://github.com/gorhill/uBlock/wiki/Blocking-mode:-hard-mode)?

**Comment by Thorin-Oakenpants** [comment]
@Atavic don't confuse your quiki's with stiki's - I answered your q in the stiki. This quiki is a one off to get my A into G and finish the wiki section1

**Comment by earthlng** [comment]
Nicely done with the Overview wiki page. I made [2 small changes](https://github.com/ghacksuserjs/ghacks-user.js/wiki/1.1-Overview/_compare/8be6aeaef1d5ce8f122faa643299dd15001636a8...d61004a3e75a10371c927f4908c356764531be7b)

teaching is a strong word, how about this instead:
* `It will teach you how to protect your privacy ...`
==> Our user.js helps to better understand how to protect ...

And there's one typo in the 2nd image: collasible

**Comment by Thorin-Oakenpants** [comment]
damn .. not sure if i can be bothered to recreate that pic

**Comment by earthlng** [comment]
np, nobody will notice it :) and remember our motto ...

**Comment by Thorin-Oakenpants** [comment]
Having issues with creating new profiles. Selecting the directory doesn't create it in that dir, but the one above it. Seems weird. Yeah, working on Implementation section.

**Comment by earthlng** [comment]
An idea: we could provide an opensearch.xml file which would search the issue tracker for a given pref.
It would be nicer if we had a logo for this user.js page but I guess we can either take your avatar or the gh favicon for now.

**Comment by earthlng** [comment]
> > And why would you want a user.js?
> It can help you to better understand how to protect your privacy ...

"a" user.js is really just a list of prefs normally, but "our" user.js can help to better understand blablabla

**Comment by Thorin-Oakenpants** [comment]
All the previous points start with "It can..", so consistency. Clearly ours is very informative, I don't think that needs to be said and it sounds a little whats-the-word .. skiteful (boastful). Any user.js can be a basis for learning, any preference could teach someone something - because people (assumption) will read what they say or experiment what they do. Ours is not exclusive in this department. Bloody awesome, but not exclusive.

Dunno why people would want to search the issue tracker (don't answer that) - the whole point is that the user.js (should) holds everything they need to know about each preference and the ramifications

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/wiki/1.3-Implementation

Finished. Implementation (and the wiki home page) are all I need to link to in the readme. The implementation page lays out all the steps and pages to read in order to use this in a sensible manner, especially for people not familiar with it.

Closing this, but post feedback if required. Now I just need to rewrite and clarify our readme: sans any technical info (that's what the wiki is for), it's more of a mission statement. It will be as short and sweet as I can get it. Thumbs up if you like my work xD
